### ISSUE TITLE
wrap code window or scroll code window

### ISSUE THREAD

**Issue by fdncred** [devyn/nu_plugin_explore_ir#1]
I was demnostrating that `$in` collects in discord using this one-liner.
```nushell
view ir { ..10 | each {|x| sleep 0.5sec; $x } | if false { where { |x| $x mod 2 == 0 } } else {$in}}
# 4 registers, 23 instructions, 0 bytes of data
   0: load-literal           %1, nothing
... blah blah blah
  17: collect                %0          # label(1), from(10:)
```
This is great but I really wanted to see the spans highlighted so I did this.
```nushell
explore ir { ..10 | each {|x| sleep 0.5sec; $x } | if false { where { |x| $x mod 2 == 0 } } else {$in}}
```
Unfortunately, if your screen isn't wide enough, you cannot see what's highlighted, nor can you scroll to the right in the `Source code` window.
![image](https://github.com/user-attachments/assets/e42e5962-e06a-4d6c-b40b-5fc688adc38f)
I had to do something like this in order to see it.
```nushell
explore ir { ..10 | each {|x| sleep 0.5sec; $x } |
  if false { where { |x| $x mod 2 == 0 } } else {$in}
}
```
![image](https://github.com/user-attachments/assets/dde22418-2cbf-46b7-865f-cf1122c57c1a)

This made me think that it would be nice if the `Source code` window would either scroll or wrap the text. If it scrolled, it would be nice if whatever was selected as made visible aka automatically scroll over to whatever needs to be highlighted.

**Comment by devyn** [comment]
Good point. Wrapping would probably be the best experience, it's just rather tough to find the span in the right place after doing some other manipulation to the text, so I need to probably set it up in some way with tuples for the original span of the text

**Comment by fdncred** [comment]
I wonder that if they're "soft wraps" i.e. there are no additional cr/lf's, maybe the span would still work with "soft wrapping". I guess it just depends on the source code widget and how it works.

**Comment by devyn** [comment]
It's pretty awfully complicated right now and I actually want to simplify it & get `nu-highlight` in there somehow

I have to split it into lines myself and then find the first and last line that have partial highlighting, as well as the lines in the middle that are fully highlighted

This is required because of the rendering model of ratatui - it's fundamentally split into spans with a particular style, grouped into lines, grouped into text.

It might be easier to just get the output of `nu-highlight` converted into styled spans on lines early, re-associate the span offsets with them (by manually counting the contents), change the style of spans in the selected span range, and then manually split lines that are too long so it can wrap

**Comment by fdncred** [comment]
I was thinking of a hack too, but slightly different and probably not a good idea since it uses regular expressions.

I took this, the output of nu-highlight
![image](https://github.com/user-attachments/assets/ce1e76a8-3273-4363-9ed6-6d2503439ac1)
and sent through debug to get the encoded ansi escape string in val.
```nushell
"..10 | each {|x| sleep 0.5sec; $x } | if false { where { |x| $x mod 2 == 0 } } else {$in}" | nu-highlight | debug -r
String {
    val: "\u{1b}[33m..\u{1b}[0m\u{1b}[1;35m10\u{1b}[0m \u{1b}[1;35m|\u{1b}[0m \u{1b}[1;36meach\u{1b}[0m \u{1b}[38;2;255;176;0m{|x\u{1b}[0m\u{1b}[1;35m|\u{1b}[0m \u{1b}[1;36msleep\u{1b}[0m \u{1b}[1;35m0.5\u{1b}[0m\u{1b}[32msec\u{1b}[0m; \u{1b}[35m$x\u{1b}[0m\u{1b}[38;2;255;176;0m }\u{1b}[0m \u{1b}[1;35m|\u{1b}[0m \u{1b}[1;36mif\u{1b}[0m \u{1b}[96mfalse\u{1b}[0m \u{1b}[38;2;51;255;0m{ \u{1b}[0m\u{1b}[1;36mwhere\u{1b}[0m \u{1b}[38;2;51;255;0m{ |x\u{1b}[0m\u{1b}[1;35m|\u{1b}[0m \u{1b}[35m$x\u{1b}[0m \u{1b}[33mmod\u{1b}[0m \u{1b}b}[1;36melse\u{1b}[0m \u{1b}[38;2;51;255;0m{\u{1b}[0m\u{1b}[35m$in\u{1b}[0m\u{1b}[38;2;51;255;0m}\u{1b}[0m",
    internal_span: Span {
        start: 375510,
        end: 375522,
    },
}
```
then parsed it with my regex plugin to find all the spans for the ansi escape sequences and ansi escapes come in 2, a prefix (color) and a suffix (reset/0m). What is between these is the text string.
![image](https://github.com/user-attachments/assets/bf402549-7a90-47b5-a992-827180aac289)
and then you can test it with `str replace`
![image](https://github.com/user-attachments/assets/00809b2b-aac4-4d1f-94fd-20e950867d49)

again, probably a bad idea but there are probably ways to parse the output of nu-highlight into something that is useable. I've also seen ansi escape parsers too, which is kind of what that tui ansi widget does I think, iirc.

The regex if you wanted to play.
```
'(?:\\u\{1b\}[@-Z\\-_]|[\x80-\x9A\x9C-\x9F]|(?:\\u\{1b\}\[|\x9B)[0-?]*[ -/]*[@-~])'
```
